src/properties/degrees.c

```diff
@@ -517,3 +517,138 @@ igraph_error_t igraph_sort_vertex_ids_by_degree(const igraph_t *graph,
+    igraph_integer_t eid;
+    igraph_integer_t v1id;
+    igraph_integer_t v2id;
+    igraph_integer_t v1deg;
+    igraph_integer_t v2deg;
+
+    if (weights && igraph_vector_int_size(weights) != no_of_edges) {
+        IGRAPH_ERRORF("Weight vector length (%" IGRAPH_PRId ") does not match number of edges (%" IGRAPH_PRId ").",
+                      IGRAPH_EINVAL,
+                      igraph_vector_int_size(weights), no_of_edges);
+    }
+
+    if (igraph_is_directed(graph)) {
+        igraph_vector_int_t out_degrees;
+        igraph_vector_int_t in_degrees;
+        igraph_integer_t max_out_degree;
+        igraph_integer_t max_in_degree;
+
+        IGRAPH_CHECK(igraph_vector_int_init(&out_degrees, 0));
+        IGRAPH_CHECK(igraph_vector_int_init(&in_degrees, 0));
+        IGRAPH_CHECK(igraph_degree(graph, &out_degrees, igraph_vss_all(), IGRAPH_OUT, true));
+        IGRAPH_CHECK(igraph_degree(graph, &in_degrees, igraph_vss_all(), IGRAPH_IN, true));
+        IGRAPH_CHECK(igraph_maxdegree(graph, &max_out_degree, igraph_vss_all(), IGRAPH_OUT, true));
+        IGRAPH_CHECK(igraph_maxdegree(graph, &max_in_degree, igraph_vss_all(), IGRAPH_IN, true));
+
+        if (dout > 0 && din > 0) {
+            if (dout < max_out_degree || din < max_in_degree) {
+                IGRAPH_ERRORF("dout (%" IGRAPH_PRId ") or din (%" IGRAPH_PRId ") are smaller than the smallest required dimensions of the matrix (%" IGRAPH_PRId ", %" IGRAPH_PRId ")",
+                              IGRAPH_EINVAL,
+                              dout, din, max_out_degree, max_in_degree);
+            }
+            IGRAPH_CHECK(igraph_matrix_int_resize(jdm, dout, din));
+        } else {
+            IGRAPH_CHECK(igraph_matrix_int_resize(jdm, max_out_degree, max_in_degree));
+        }
+        // Set all elements to 0
+        igraph_matrix_int_null(jdm);
+
+        IGRAPH_CHECK(igraph_es_all(&es, IGRAPH_EDGEORDER_ID));
+        IGRAPH_CHECK(igraph_eit_create(graph, es, &eit));
+
+        while (!IGRAPH_EIT_END(eit)) {
+            eid = IGRAPH_EIT_GET(eit);
+            v1id = IGRAPH_FROM(graph, eid);
+            v2id = IGRAPH_TO(graph, eid);
+            v1deg = igraph_vector_int_get(&out_degrees,v1id);
+            v2deg = igraph_vector_int_get(&in_degrees, v2id);
+            if (!weights) {
+                MATRIX(*jdm, v1deg-1, v2deg-1)++;
+            } else {
+                MATRIX(*jdm, v1deg-1, v2deg-1) += VECTOR(*weights)[eid];
+            }
+            IGRAPH_EIT_NEXT(eit);
+        }
+
+        igraph_vector_int_destroy(&out_degrees);
+        igraph_vector_int_destroy(&in_degrees);
+
+    } else {
+        igraph_vector_int_t degrees;
+        igraph_integer_t max_degree;
+        // Compute max degrees
+        IGRAPH_CHECK(igraph_vector_int_init(&degrees, 0));
+        IGRAPH_CHECK(igraph_degree(graph, &degrees, igraph_vss_all(), IGRAPH_ALL, true));
+        IGRAPH_CHECK(igraph_maxdegree(graph, &max_degree, igraph_vss_all(), IGRAPH_ALL, true));
```

Comment:
Instead of re-computing degrees again, we can do a small optimization here and take the max of the existing degree vector using `igraph_vector_int_max()`. This function must not be called for empty vectors, so let us check `no_of_nodes` and if it is zero, do not call it (just use `0` as the "maximum").